File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();
        char[] chars = s.toCharArray();
        String str = "";
        int start = 0;
        int maxLength = 0;
        for (int i = 0; i < chars.length; i++) {
            if (map.containsKey(chars[i])) {
                int pos = map.get(chars[i]);
                if (start <= pos) {
                    start = pos + 1;
                }
                str = str + chars[i];
                map.put(chars[i], str.length() - 1);
            } else {
                str = str + chars[i];
                map.put(chars[i], str.length() - 1);
            }
            maxLength = Math.max(maxLength, str.length() - start);
        }
        return maxLength;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> hashMap;
        int n = s.size();
        int left = 0;
        int right = 0;
        int len = 0;
        int maxLen = 0;
        while (right < n) {
            if (hashMap.find(s[right]) != hashMap.end()) {
                if (hashMap[s[right]] >= left) {
                    left = hashMap[s[right]] + 1;
                }
            }
            hashMap[s[right]] = right;
            right++;
            maxLen = max(maxLen, right - left);
        }
        return maxLen;
    }
};